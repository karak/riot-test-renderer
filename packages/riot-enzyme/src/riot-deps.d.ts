declare module 'riot-compiler' {
  function compile(src: string, userOpts?: any): string;
}

declare module 'riot-tmpl' {
  function tmpl(template: string, data: any): any;
  const brackets: any;
}
