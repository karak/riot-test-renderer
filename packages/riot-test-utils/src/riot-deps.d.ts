declare module 'riot-compiler' {
  function compile(src: string, userOpts?: any): string;
}

declare module 'riot-tmpl' {
  function tmpl(template: string, data: any): any;
  const brackets: any;
}

declare module 'riot-observable' {
  import { observable, ObservableCallback } from 'riot';
  export default observable;
  export { ObservableCallback };
}
