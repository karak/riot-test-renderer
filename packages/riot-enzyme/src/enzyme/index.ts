import {
  configure,
  shallow as reactShallow,
  ShallowRendererProps as ReactShallowRendererProps,
  ShallowWrapper as ReactShallowWrapper,
  EnzymePropSelector,
} from 'enzyme';
import { TagInstance } from 'riot';

import EnzymeRiotAdapter from './adapter';
import RiotShallowRendererProps from './RiotShallowRendererProps';
import toJSON from 'riot-test-utils/dist/lib/toJSON';
import isString from 'lodash/isString';

configure({ adapter: new EnzymeRiotAdapter() });

const SINGLE_TAG_NAME_REGEX = /^[^<]*<(.*?)>/m;

export interface TagOpts {
  [name: string]: any;
}

/**
 * Guess its name from single tag source
 * without real compilation
 *
 * @param singleSource tag source which may be contain single tag
 * @returns tag name
 * @throws {Error} if no tag expression is found
 */
function guessNameFromSource(singleSource: string): string {
  const m = SINGLE_TAG_NAME_REGEX.exec(singleSource);
  if (m === null) throw new Error('No tag expression with <>');

  return m[1];
}

/**
 * Shallow render
 *
 * @param singleSource tag source which contains single tag
 * @param opts tag interface
 * @param options renderer options see {@see enzyme~shallow}
 * @returns wrapper object of rendered element.
 */
export function shallow<TOpts extends TagOpts>(
  singleSource: string,
  opts?: TOpts,
  options?: ReactShallowRendererProps
): ShallowWrapper<TOpts>;
/**
 * Shallow render
 *
 * @param source tag source
 * @param name tag name
 * @param opts tag interface
 * @param options renderer options see {@see enzyme~shallow}
 * @returns wrapper object of rendered element.
 */
export function shallow<TOpts extends TagOpts>(
  source: string,
  name: string,
  opts?: TOpts,
  options?: ReactShallowRendererProps
): ShallowWrapper<TOpts | {}>;
export function shallow<TOpts extends TagOpts>(
  source: string,
  ...args: any[]
): ShallowWrapper<TOpts> {
  let name: string;
  let opts: TOpts | undefined;
  let options: ReactShallowRendererProps | undefined;

  if (isString(args[0])) {
    /* Case: source, name, opts?, options?*/
    name = args[0];
    opts = args[1];
    // tslint:disable-next-line:no-magic-numbers
    options = args[2];
  } else {
    /* Case: source, opts?, options?*/
    name = guessNameFromSource(source);
    // tslint:disable-next-line:no-magic-numbers
    opts = args[0];
    // tslint:disable-next-line:no-magic-numbers
    options = args[1];
  }

  const element = {
    type: name,
    props: opts,
    key: null,
  };
  const adaptee = reactShallow(element, {
    'riot-enzyme': { source },
    ...options,
  } as RiotShallowRendererProps);
  return new ShallowWrapper(adaptee);
}

/**
 * Wrapper of shallow-rendered element
 * like {@see enzyme~ShallowWrapper}
 */
export class ShallowWrapper<TOpts extends TagOpts = any> {
  constructor(private adaptee: ReactShallowWrapper<TOpts | undefined>) {}

  unmount() {
    this.adaptee.unmount();
  }

  name(): string {
    return this.adaptee.type() as string;
  }

  opts(): TOpts | {} {
    return this.adaptee.props() || {};
  }

  find(props: EnzymePropSelector): ShallowWrapper<any>;
  find(selector: string): ShallowWrapper<any>;
  find(selector: any) {
    const nextAdaptee = this.adaptee.find(selector) as ReactShallowWrapper<any>;
    return new ShallowWrapper(nextAdaptee);
  }

  at(index: number) {
    return new ShallowWrapper(this.adaptee.at(index));
  }

  text() {
    return this.adaptee.text();
  }

  html() {
    return this.adaptee.html();
  }

  toJson(): object | null {
    const root = ((this.adaptee.instance() as any) as TagInstance).root;
    return toJSON(root !== undefined ? root : (null as any)) as object | null;
  }
}
