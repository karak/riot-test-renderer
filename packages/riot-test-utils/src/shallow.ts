import { TagInstance } from 'riot';
import ShallowWrapper from './lib/ShallowWrapper';
import RiotShallowRenderer from './lib/RiotShallowRenderer';
import shallow from './lib/shallow';

export { TagInstance, ShallowWrapper, RiotShallowRenderer, shallow };
