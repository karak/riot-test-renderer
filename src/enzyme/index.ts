import {
  configure,
  shallow as reactShallow,
  ShallowRendererProps as ReactShallowRendererProps,
  ShallowWrapper as ReactShallowWrapper,
  EnzymePropSelector,
} from 'enzyme';

import EnzymeRiotAdapter from './adapter';
import RiotShallowRendererProps from './RiotShallowRendererProps';
import TagInstance from '../lib/TagInstance';
import toJSON from '../lib/toJSON';
import isString from 'lodash/isString';

configure({ adapter: new EnzymeRiotAdapter() });

const SINGLE_TAG_NAME_REGEX = /^[^<]*<(.*?)>/m;

/**
 * Guess its name from single tag source
 * without real compilation
 *
 * @param singleSource tag source which may be contain single tag
 * @returns tag name
 * @throws {Error} if no tag expression is found
 */
function guessNameFromSource(singleSource: string): string {
  const m = SINGLE_TAG_NAME_REGEX.exec(singleSource);
  if (m === null) throw new Error('No tag expression with <>');

  return m[1];
}

/**
 * Shallow render
 *
 * @param singleSource tag source which contains single tag
 * @param opts tag interface
 * @param options renderer options see {@see enzyme~shallow}
 * @returns wrapper object of rendered element.
 */
export function shallow<TOpts>(
  singleSource: string,
  opts?: TOpts,
  options?: ReactShallowRendererProps
): ShallowWrapper<TOpts>;
/**
 * Shallow render
 *
 * @param source tag source
 * @param name tag name
 * @param opts tag interface
 * @param options renderer options see {@see enzyme~shallow}
 * @returns wrapper object of rendered element.
 */
export function shallow<TOpts>(
  source: string,
  name: string,
  opts?: TOpts,
  options?: ReactShallowRendererProps
): ShallowWrapper<TOpts>;
export function shallow<TOpts>(
  source: string,
  ...args: any[]
): ShallowWrapper<TOpts> {
  let name: string;
  let opts: TOpts | undefined;
  let options: ReactShallowRendererProps | undefined;

  if (isString(args[0])) {
    name = args[0];
    opts = args[1];
    // tslint:disable-next-line:no-magic-numbers
    options = args[2];
  } else {
    name = guessNameFromSource(source);
    // tslint:disable-next-line:no-magic-numbers
    opts = args[2];
    // tslint:disable-next-line:no-magic-numbers
    options = args[3];
  }

  const adaptee = reactShallow({ type: name, props: opts || {}, key: null }, {
    'riot-enzyme': { source },
    ...options,
  } as RiotShallowRendererProps);
  return new ShallowWrapper(adaptee);
}

/**
 * Wrapper of shallow-rendered element
 * like {@see enzyme~ShallowWrapper}
 */
export class ShallowWrapper<TOpts = any> {
  constructor(private adaptee: ReactShallowWrapper) {}

  unmount() {
    this.adaptee.unmount();
  }

  name(): string {
    return this.adaptee.type() as string;
  }

  opts(): TOpts {
    return this.adaptee.props() as TOpts;
  }

  find(props: EnzymePropSelector): ShallowWrapper<any>;
  find(selector: string): ShallowWrapper<any>;
  find(selector: any) {
    const nextAdaptee = this.adaptee.find(selector);
    return new ShallowWrapper(nextAdaptee);
  }

  at(index: number) {
    return new ShallowWrapper(this.adaptee.at(index));
  }

  text() {
    return this.adaptee.text();
  }

  html() {
    return this.adaptee.html();
  }

  toJson(): object | null {
    const root = ((this.adaptee.instance() as any) as TagInstance).root;
    return toJSON(root !== undefined ? root : (null as any)) as object | null;
  }
}
