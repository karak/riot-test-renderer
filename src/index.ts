export * from './shallow';
export * from './vdom';
