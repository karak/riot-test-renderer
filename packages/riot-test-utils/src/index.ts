/* Riot dependency in public APIs */
import { TagInstance, TagOpts, TagRefs, NestedTags } from 'riot';
export { TagInstance, TagOpts, TagRefs, NestedTags };

/* High level APIs */
import shallow from './lib/wrappers/shallow';
import ShallowWrapper from './lib/wrappers/ShallowWrapper';
export { ShallowWrapper, shallow };
export * from './lib/wrappers/mount'; // !!now developing!!

/* Lowlevel APIs */
export * from './lib/renderers';
import Simulate from './lib/Simulate';
export { Simulate };
export * from './lib/transform';
