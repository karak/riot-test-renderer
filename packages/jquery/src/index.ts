import { extend, RiotWrapper, WeakWrapper } from 'riot-test-utils';
import { default as $ } from 'jquery';

function isRiotWrapper(
  wrapper: RiotWrapper | WeakWrapper
): wrapper is RiotWrapper {
  return wrapper.instance !== null;
}

function toJQuery(wrapper: RiotWrapper | WeakWrapper): JQuery {
  return $(isRiotWrapper(wrapper) ? wrapper.root : wrapper.get());
}

/** An extension methods with jQuery */
const jqueryExtension: {
  [name: string]: Function;
} = {};

/** Add the method to get jQuery object($()) */
function addWrap() {
  jqueryExtension['$'] = function(this: RiotWrapper | WeakWrapper) {
    return toJQuery(this);
  };
}

/** Add a method of query that returns boolean value */
function addQuery(name: string) {
  jqueryExtension[name] = function(this: RiotWrapper | WeakWrapper) {
    const $el = toJQuery(this);
    return ($el as any)[name].apply($el, arguments);
  };
}

/** Add a method of accessor */
function addAccessor(name: string, isSetter = isSetterDefault) {
  jqueryExtension[name] = function(this: RiotWrapper | WeakWrapper) {
    const $el = toJQuery(this);
    if (!isSetter.apply(this, arguments)) {
      return ($el as any)[name].apply($el, arguments);
    } else {
      ($el as any)[name].apply($el, arguments);
      return this;
    }
  };
}

function isSetterDefault(this: JQuery) {
  return arguments.length > 1;
}

/* Get jQuery itself */
addWrap();

/* Assertion helpers */
addQuery('is');
addQuery('hasClass');
addAccessor('attr');
addAccessor('css', function() {
  return arguments.length > 1 || typeof arguments[0] === 'object';
});
addAccessor('data');
addAccessor('prop');
addAccessor('value');

/* Do extend */
extend(jqueryExtension);

/* TODO: DOM manipulation methods like filter, has, find, children */

/* Modify type definition */
declare module 'riot-test-utils' {
  interface WrapperExtensions {
    $(): JQuery;
    is: typeof $.prototype.is;
    hasClass: typeof $.prototype.hasClass;
    attr: typeof $.prototype.attr;
    css: typeof $.prototype.css;
    data: typeof $.prototype.data;
    prop: typeof $.prototype.prop;
    value: typeof $.prototype.value;
  }
}
