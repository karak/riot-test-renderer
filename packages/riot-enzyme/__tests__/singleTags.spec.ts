import { shallow, ShallowWrapper } from '../src';
import {
  staticTag,
  tagWithOpts,
} from 'riot-test-renderer-shared/tags/singleTags';

describe('enzyme-riot-adapter', () => {
  describe('staticTag', () => {
    let wrapper: ShallowWrapper<{}>;

    beforeEach(() => {
      wrapper = shallow(staticTag, 'static');
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it('renders html', () => {
      expect(wrapper.html()).toBe(
        '<static data-is="static"><p>Hello, world!</p></static>'
      );
    });

    it('finds by name', () => {
      expect(wrapper.find('p').text()).toBe('Hello, world!');
    });

    it('has empty opts from undefined', () => {
      expect(wrapper.opts()).toEqual({});
    });

    it('has a JSON expression', () => {
      expect(wrapper.toJson()).toEqual({
        name: 'static',
        attributes: { 'data-is': ['static'] },
        children: [
          {
            name: 'p',
            attributes: {},
            children: ['Hello, world!'],
          },
        ],
      });
    });
  });

  describe('tagWithOpts', () => {
    let wrapper: ShallowWrapper<{}>;

    beforeEach(() => {
      wrapper = shallow(tagWithOpts, 'tag', { data: 'Hello, world!' });
    });

    afterEach(() => {
      wrapper.unmount();
    });

    it('shoud have the same opts as passed', () => {
      expect(wrapper.opts()).toEqual({ data: 'Hello, world!' });
    });

    it('should have a JSON expression', () => {
      expect(wrapper.toJson()).toEqual({
        name: 'tag',
        attributes: { 'data-is': ['tag'] },
        children: [
          {
            name: 'p',
            attributes: {},
            children: ['Hello, world!'],
          },
        ],
      });
    });

    it('should match snapshot', () => {
      expect(wrapper.toJson()).toMatchSnapshot();
    });
  });
});
