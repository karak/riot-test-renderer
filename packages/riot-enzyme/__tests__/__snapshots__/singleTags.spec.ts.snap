// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme-riot-adapter tagWithOpts should match snapshot 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        "Hello, world!",
      ],
      "name": "p",
      "opts": Object {},
    },
  ],
  "name": "tag",
  "opts": Object {},
}
`;
